### ISSUE TITLE
[Paid: Final Steps] [$1000] Web - Chat - When copying part of a multiline message, the entire message is copied

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#17868]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open the ND on the mac/Safari
2. Send multiline message
3. Select one word of the msg, click on the selected text via right button: context menu is opened
4. Click on the Copy icon
5. Paste the text

## Expected Result:
Only selected text should be copied

## Actual Result:
The whole message has been copied at the Safari

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.4.0

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/233864487-1eebda4e-f80d-4e33-b3b8-2b80617cab7c.mp4

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01075913f2cc61e719</li>
        <li>Upwork Job ID: 1650506933540429824</li>
        <li>Last Price Increase: 2023-05-08</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @conorpendergrast (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by tienifr** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

In Safari, when we try to select a part of a multiline message, then right click the selected text and press `Copy to clipboard`, it copies the whole message instead of the selected text

### What is the root cause of that problem?

Let's look at the flow of the [executeSecondaryInteractionOnContextMenu()](https://github.com/Expensify/App/blob/931e15f7d7728f7be8d4f6190d8e9a2ffc6b5dac/src/components/PressableWithSecondaryInteraction/index.js#L32-L49) function, which would be triggered when we right click a message. Basically, the current logic is `blurring all active components -> showing the popover menu`

https://github.com/Expensify/App/blob/931e15f7d7728f7be8d4f6190d8e9a2ffc6b5dac/src/components/PressableWithSecondaryInteraction/index.js#L38-L48

The inconsistency lies in this piece of code. The `blur()` function has different behavior on Safari and Chrome. On Safari, the `blur()` function would de-select the user-selected text, causing the [window.getSelection()](https://github.com/Expensify/App/blob/137b0e87aaf3c737ce64befe417dda08c3ff6320/src/libs/SelectionScraper/index.js#L18) function to return null, while it would not on Chrome. When the line `this.pressableRef.blur()` is triggered, the Safari browser would de-select the selected text, affecting the [logic](https://github.com/Expensify/App/blob/137b0e87aaf3c737ce64befe417dda08c3ff6320/src/libs/SelectionScraper/index.js#L18) that was used to get the current selected text. Here, because the text was de-selected, and we are opening the popover **after blurring**, the `window.getSelection()` function would return a null component. The copied text now would be the whole message, which is similar to the behavior when we right click on the message without selecting anything.

In short, after blurring the text, the selection on Safari is cleared, while it is not on Chrome. Therefore, when the `showPopover()` function is triggered afterwards, the selected component is null on Safari, which make the app understand that the whole long message should be copied.

I believe this is a regression from https://github.com/Expensify/App/pull/15375, where the blurring logic was added. A very similar issue of the selection API inconsistency after a component is blurred on Chrome and Safari (in another repo) can be found [here](https://github.com/ianstormtaylor/slate/issues/4376). A possibility is that there is a difference between the way Safari and Chrome handle the blurring event.

### What changes do you think we should make in order to solve the problem?

Changing the order from `blurring -> showPopover` to `showPopover -> blurring` will solve this issue

```diff   
-        if (this.props.withoutFocusOnSecondaryInteraction && this.pressableRef) {
-           this.pressableRef.blur();
-        }
-        this.props.onSecondaryInteraction(e);
+        this.props.onSecondaryInteraction(e);
+        if (this.props.withoutFocusOnSecondaryInteraction && this.pressableRef) {
+           this.pressableRef.blur();
+        }
      
```

I think this is a reasonable change, since 
1. It's minimal. The changes is small.
2. It's logical. The `onSecondaryInteraction()` function is now using the user-selected text before blurring the current element, so we don't have to worry about any unexpected behavior that affect the user-selected text.
3. It will not re-introduce the bug that was fixed in https://github.com/Expensify/App/pull/15375.

### Alternate solutions

Alternatively, we could add a `setTimeout` around the `blur()` function. IMO this is a workaround and not a good way to do it, but given the `blur()` function has some kind of a small delay, and we want to preserve the selected component when the `onSecondaryInteraction()` is called, this is worth mentioning.
```diff
+            setTimeout(() => {
                this.pressableRef.blur();
+            }, 0)
```

### Records
Working well after the fix

https://user-images.githubusercontent.com/113963320/233944342-8afba897-0d59-4f4f-b29a-529cda7a3d86.mov

**Comment by conorpendergrast** [comment]
This isn't a bug: the context menu applies to all parts of the message, not individual components within the message. 
eg: you can't select part of the message and delete just part of the message, or mark just part of the message as unread.

Closing as it's not a bug!

**Comment by tienifr** [comment]
@conorpendergrast 
I don't agree with it. Even if it's not a bug, the behavior is inconsistent in different browsers. On Chrome or Desktop, selecting a part of the multiline text then right click -> copy would copy the selected only. Meanwhile in Safari, it would copy the whole text.

**Comment by conorpendergrast** [comment]
Oh that's very weird - I didn't spot that this was inconsistent in the OP (there's no platforms selected)

**Comment by conorpendergrast** [comment]
Reopening as this is an inconsistency that we should resolve either as copying the full message, or copying the selected part of the message

**Comment by conorpendergrast** [comment]
Updated the OP to include the platform!

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @fedirjh (`External`)

**Comment by tienifr** [comment]
@fedirjh Thank you for the response, that’s a nice catch!

I’m not sure with the underlying principle of how Safari handle the blurring event, but after some more careful examination, here is my observation:

When we are copying a normal text on Safari, the flow would be
- Select text -> right click -> **selection got cleared** -> modal state is updated

In this situation, the modal `selection` value inside `showContextMenu()` function is `‘’` because the selection is already cleared when it reach that point, which will make it copy the whole message.

On the other hand, when we are copying a codeblock text, the flow would be
- Select text -> right click -> modal state is updated -> **selection got cleared**

In both of the situations, the `window.getSelection()` after everything is executed would be null.
<img width="830" alt="image" src="https://user-images.githubusercontent.com/113963320/234466108-f369456a-6a1d-4259-892f-30048f13557f.png">



Meanwhile, the `window.getSelection()` value is not cleared on Chrome
<img width="539" alt="image" src="https://user-images.githubusercontent.com/113963320/234465651-c14f0635-773d-4a0c-bbf4-47de0a8be0cc.png">

This is only happening of Safari, so IMO the root cause lay in some difference in the way Safari handle the blurring action, which may affect the order of the actual actions being executed. The order is different between when we select a pure text and when we select a code block.

**Comment by tienifr** [comment]
@fedirjh Thank you for the input, and also the very helpful link!

I can confirm that this is happening on Safari (and not on Chrome). The copied text would be a single word that the pointer was hovered in when the right click event is triggered. I'm taking a deeper look into this.

**Comment by tienifr** [comment]
Apparently, the `window.getSelection()` result (which contains startOffset, startContainer, endOffset, endContainer) is different on Safari when we select across multiple messages (let's say message1,message2 for example) and right click. On Chrome, the startOffset and startContainer corresponds to message1, endOffset and endContainer correspond to message2. While on Safari, both start and end value for offset and container correspond to the one word that was clicked.

Still finding a proper way to address this.

**Comment by jasperhuangg** [comment]
@tienifr is very close to finishing his proposal, he just has a few edge cases to iron out.

**Comment by jasperhuangg** [comment]
Still waiting on proposals

**Comment by conorpendergrast** [comment]
@tienifr Do you still have a proposal in the works for this?

**Comment by tienifr** [comment]
@fedirjh @conorpendergrast I'm still trying to fix the [edge cases](https://github.com/Expensify/App/issues/17868#issuecomment-1528153080)

**Comment by tienifr** [comment]
After some investigation, I am still not able to find a resolution for this. The issue remains in the way Safari handle selection, which I have [explained](https://github.com/Expensify/App/issues/17868#issuecomment-1529797645) earlier. 

I did some further digging on other messaging apps, and here is my sum up.

**Action**: Select text across 2 different messages -> Right click -> Copy text. Notice that we are selecting across different messages, not across lines in a single message.

**Behavior of my proposal**
+ Chrome: the selected is copied (expected)
+ Safari: the copied text is a single word being hovered before the right click event (*)

**Behavior of other messaging apps**
- Skype
  + Chrome: the selected is copied (expected)
  + Safari: the whole *later* message is copied, which does not include the text from the former text

- Slack
  + On Chrome: The text is copied, but with extra time information (not expected)
  + On Safari: same as (*)

- Linkedin
  + On Chrome: The text is copied, but with extra emojis from reaction menu (not expected)
  + On Safari: same as (*)

- Messenger
  + On Chrome: the text is copied, but with an extra “You have sent” text in the middle (not expected)
  + On Safari: same as (*)

Given that, I think this edge case should be addressed on the Webkit side, or on another issue, and it is out of scope in the current context (copying multiline text).

Thoughts?

**Comment by conorpendergrast** [comment]
@jasperhuangg Waiting for your review to confirm and assign @tienifr

**Comment by conorpendergrast** [comment]
Offers sent via Upwork, will pay after the regression wait period

**Comment by conorpendergrast** [comment]
Pay day, moving to Daily

**Comment by conorpendergrast** [comment]
Paid via Upwork. 

@fedirjh Final step is [the checklist](https://github.com/Expensify/App/issues/17868#issuecomment-1550461887)!

**Comment by conorpendergrast** [comment]
I agree with those reproduction steps. Created [the regression test request](https://github.com/Expensify/Expensify/issues/285701), and all done!
